// @dynamic-labs/utils/src/setupInsideIframe/utils/setupPlatformService/setupPlatformService.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../_virtual/_tslib.cjs');
var createMessageSender = require('../createMessageSender/createMessageSender.cjs');
var createMessageHandler = require('../createMessageHandler/createMessageHandler.cjs');
var PlatformService = require('../../../services/PlatformService/PlatformService.cjs');
require('tldts');

const setupPlatformService = (initialParentURL) => {
    let parentURL = new URL(initialParentURL);
    const messageHandler = createMessageHandler.createMessageHandler(initialParentURL.origin);
    const messageSender = createMessageSender.createMessageSender(initialParentURL.origin);
    PlatformService.PlatformService.implementation = {
        downloadFile: (fileName, file) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
            const element = document.createElement('a');
            element.href = URL.createObjectURL(file);
            element.download = fileName;
            document.body.appendChild(element);
            element.click();
            document.body.removeChild(element);
            URL.revokeObjectURL(element.href);
        }),
        getDisplayOrigin: () => parentURL.origin,
        getHost: () => parentURL.host,
        getHostname: () => parentURL.hostname,
        getOrigin: () => parentURL.origin,
        getTLD: () => parentURL.hostname,
        getUrl: () => parentURL,
        isNativeMobile: false,
        openURL: (url_1, ...args_1) => _tslib.__awaiter(void 0, [url_1, ...args_1], void 0, function* (url, target = 'self', features = '') {
            return messageSender('OPEN_URL', {
                features,
                target,
                url,
            });
        }),
    };
    messageHandler('UPDATE_PARENT_URL', ({ url }) => {
        parentURL = new URL(url);
    });
};

exports.setupPlatformService = setupPlatformService;
