// @dynamic-labs-sdk/client/dist/getSignedSessionId-QJpJQ_Jz.cjs.js
const require_constants = require('./constants-BbG0v26E.cjs.js');
const require_isMfaRequiredForAction = require('./isMfaRequiredForAction-CAwdzRA9.cjs.js');

//#region src/modules/mfa/consumeMfaTokenIfRequiredForAction/consumeMfaTokenIfRequiredForAction.ts
const consumeMfaTokenIfRequiredForAction = async ({ mfaAction }, client) => {
	if (!await require_isMfaRequiredForAction.isMfaRequiredForAction({ mfaAction }, client)) return;
	return require_isMfaRequiredForAction.consumeMfaToken(client);
};

//#endregion
//#region src/utils/p256Sign/p256Sign.ts
const p256Sign = async ({ privateKeyJwk, message }) => {
	const privateKey = await crypto.subtle.importKey("jwk", privateKeyJwk, {
		name: "ECDSA",
		namedCurve: "P-256"
	}, false, ["sign"]);
	return require_isMfaRequiredForAction.bufferToHex(await crypto.subtle.sign({
		hash: "SHA-256",
		name: "ECDSA"
	}, privateKey, new TextEncoder().encode(message)));
};

//#endregion
//#region src/modules/sessionKeys/generateSessionSignature/generateSessionSignature.ts
const generateSessionSignature = async ({ sessionId }, client) => {
	const sessionKeys = require_constants.getSessionKeys(client);
	require_constants.assertDefined(sessionKeys, "Session keys not found");
	const { privateKeyJwk } = sessionKeys;
	return { signature: await p256Sign({
		message: sessionId,
		privateKeyJwk
	}) };
};

//#endregion
//#region src/modules/sessionKeys/getSignedSessionId/getSignedSessionId.ts
const getSignedSessionId = async (client) => {
	const sessionId = require_constants.getCore(client).state.get().user?.sessionId;
	require_constants.assertDefined(sessionId, "Session ID is required");
	const { signature } = await generateSessionSignature({ sessionId }, client);
	require_constants.assertDefined(signature, "Signed session ID is required");
	return { signature };
};

//#endregion
Object.defineProperty(exports, 'consumeMfaTokenIfRequiredForAction', {
  enumerable: true,
  get: function () {
    return consumeMfaTokenIfRequiredForAction;
  }
});
Object.defineProperty(exports, 'getSignedSessionId', {
  enumerable: true,
  get: function () {
    return getSignedSessionId;
  }
});
