// @dynamic-labs-sdk/client/dist/isMfaRequiredForAction-CAwdzRA9.cjs.js
const require_constants = require('./constants-BbG0v26E.cjs.js');

//#region src/utils/bufferToHex/bufferToHex.ts
/**
* Converts an ArrayBuffer to a hex-encoded string
*/
const bufferToHex = (buffer) => [...new Uint8Array(buffer)].map((x) => x.toString(16).padStart(2, "0")).join("");

//#endregion
//#region src/modules/mfa/consumeMfaToken/consumeMfaToken.ts
/**
* Consumes and clears the current MFA token from the client state.
*
* This function retrieves the MFA token obtained from a successful MFA authentication,
* and removes it from the client, to prevent it from being used again.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns The MFA token that was consumed.
* @throws Error if no MFA token is found.
*/
const consumeMfaToken = (client = require_constants.getDefaultClient()) => {
	const core = require_constants.getCore(client);
	const mfaToken = core.state.get().mfaToken;
	require_constants.assertDefined(mfaToken, "No MFA token found");
	core.state.set({ mfaToken: null });
	return mfaToken;
};

//#endregion
//#region src/modules/mfa/getMfaMethods/getMfaMethods.ts
/**
* Retrieves the available MFA methods for the current user.
*
* This function fetches information about which multi-factor authentication
* methods are available and configured for the user's account.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves to the user's MFA methods configuration.
*/
const getMfaMethods = async (client = require_constants.getDefaultClient()) => {
	const core = require_constants.getCore(client);
	return await require_constants.createApiClient({}, client).getUserMfaMethods({
		environmentId: core.environmentId,
		verifiedOnly: true
	});
};

//#endregion
//#region src/modules/mfa/isMfaRequiredForAction/isMfaRequiredForAction.ts
/**
* Checks if MFA is required for a specific action.
*
* This function determines whether multi-factor authentication is required
* for the specified action based on project settings and user configuration.
*
* @param params.mfaAction - The action to check MFA requirements for.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves to true if MFA is required for the action, false otherwise.
*/
const isMfaRequiredForAction = async ({ mfaAction }, client = require_constants.getDefaultClient()) => {
	const projectSettings = client.projectSettings;
	if (!projectSettings?.security?.mfa?.actions?.some((action) => action.action === mfaAction && action.required)) return false;
	if (projectSettings?.security?.mfa?.required) return true;
	return (await getMfaMethods(client)).userHasVerifiedMfaMethods;
};

//#endregion
Object.defineProperty(exports, 'bufferToHex', {
  enumerable: true,
  get: function () {
    return bufferToHex;
  }
});
Object.defineProperty(exports, 'consumeMfaToken', {
  enumerable: true,
  get: function () {
    return consumeMfaToken;
  }
});
Object.defineProperty(exports, 'getMfaMethods', {
  enumerable: true,
  get: function () {
    return getMfaMethods;
  }
});
Object.defineProperty(exports, 'isMfaRequiredForAction', {
  enumerable: true,
  get: function () {
    return isMfaRequiredForAction;
  }
});
