// @dynamic-labs-sdk/client/dist/NotWaasWalletAccountError-DUvIdb2w.cjs.js
const require_constants = require('./constants-BbG0v26E.cjs.js');
const require_getVerifiedCredentialForWalletAccount = require('./getVerifiedCredentialForWalletAccount-DAb9lFwa.cjs.js');

//#region src/modules/user/refreshAuth/refreshAuth.ts
/**
* Refreshes the current user's data from the server.
*
* This function fetches the latest user information and token from the backend
* updating the local user and token states with any changes.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves to the verify response.
*/
const refreshAuth = async (client = require_constants.getDefaultClient()) => {
	const core = require_constants.getCore(client);
	const response = await require_constants.createApiClient({}, client).refreshAuth({ environmentId: core.environmentId });
	require_getVerifiedCredentialForWalletAccount.updateAuthFromVerifyResponse({ response }, client);
	return response;
};

//#endregion
//#region src/errors/NotWaasWalletAccountError.ts
var NotWaasWalletAccountError = class extends require_constants.BaseError {
	constructor({ walletAddress }) {
		super({
			cause: null,
			code: "not_waas_wallet_account_error",
			docsUrl: null,
			name: "NotWaasWalletAccountError",
			shortMessage: `Wallet account ${walletAddress} is not a Dynamic WaaS wallet account`
		});
	}
};

//#endregion
Object.defineProperty(exports, 'NotWaasWalletAccountError', {
  enumerable: true,
  get: function () {
    return NotWaasWalletAccountError;
  }
});
Object.defineProperty(exports, 'refreshAuth', {
  enumerable: true,
  get: function () {
    return refreshAuth;
  }
});
