// @dynamic-labs-sdk/client/dist/waasCore.cjs.js
const require_getNetworkProviderFromNetworkId = require('./getNetworkProviderFromNetworkId-CTKZxd1T.cjs.js');
const require_constants = require('./constants-BbG0v26E.cjs.js');
require('./isMfaRequiredForAction-CAwdzRA9.cjs.js');
const require_InvalidParamError = require('./InvalidParamError-C5YtT7LQ.cjs.js');
const require_getSignedSessionId = require('./getSignedSessionId-QJpJQ_Jz.cjs.js');
let _dynamic_labs_sdk_assert_package_version = require("@dynamic-labs-sdk/assert-package-version");
let _dynamic_labs_sdk_api_core = require("@dynamic-labs/sdk-api-core");
let _dynamic_labs_wallet_browser_wallet_client = require("@dynamic-labs-wallet/browser-wallet-client");

//#region src/modules/waas/createWaasClient/createWaasClient.ts
const createWaasClient = ({ chainName }, client) => {
	const core = require_constants.getCore(client);
	return new _dynamic_labs_wallet_browser_wallet_client.DynamicWalletClient({
		authMode: require_constants.isCookieEnabled(client) ? "cookie" : "header",
		authToken: client.token || "",
		baseApiUrl: (core.apiBaseUrl || require_constants.DEFAULT_WAAS_BASE_API_URL).replace("/api/v0", ""),
		baseMPCRelayApiUrl: client.projectSettings?.sdk.waas?.relayUrl || require_constants.DEFAULT_WAAS_BASE_MPC_RELAY_API_URL,
		chainName,
		environmentId: core.environmentId,
		sdkVersion: `${require_constants.CLIENT_SDK_NAME}/${core.version}`
	});
};

//#endregion
//#region src/modules/waas/getWaasChainNameFromChain/getWaasChainNameFromChain.ts
const getWaasChainNameFromChain = (chain) => {
	return require_constants.CHAINS_INFO_MAP[chain].waasChainNameOverride || chain;
};

//#endregion
//#region src/modules/waas/createWaasProvider/createWaasProvider.ts
const RAW_MESSAGE_MESSAGE_REQUIRED_LENGTH = 64;
const createWaasProvider = ({ sdkClient, chain }) => {
	const logger = require_constants.getCore(sdkClient).logger;
	let _waasClient;
	const getWaasClient = async () => {
		if (!_waasClient) {
			_waasClient = createWaasClient({ chainName: getWaasChainNameFromChain(chain) }, sdkClient);
			await _waasClient.initialize();
		}
		return _waasClient;
	};
	const backupKeySharesToGoogleDrive = async ({ password, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		return (await getWaasClient()).backupKeySharesToGoogleDrive({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			password,
			signedSessionId
		});
	};
	const createWalletAccount = async (args = {}) => {
		logger.debug("[createWaasProvider] createWalletAccount", { thresholdSignatureScheme: args.thresholdSignatureScheme });
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		return await (await getWaasClient()).createWalletAccount({
			authToken: sdkClient.token ?? void 0,
			password: args.password,
			signedSessionId,
			thresholdSignatureScheme: args.thresholdSignatureScheme ?? _dynamic_labs_wallet_browser_wallet_client.ThresholdSignatureScheme.TWO_OF_TWO
		});
	};
	const delegateKeyShares = async ({ password, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		return (await getWaasClient()).delegateKeyShares({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			password,
			signedSessionId
		});
	};
	const exportClientKeyshares = async ({ password, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		await (await getWaasClient()).exportClientKeyshares({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			password,
			signedSessionId
		});
	};
	const exportPrivateKey = async ({ displayContainer, walletAccount, password }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		const waasClient = await getWaasClient();
		const mfaToken = await require_getSignedSessionId.consumeMfaTokenIfRequiredForAction({ mfaAction: _dynamic_labs_sdk_api_core.MFAAction.WalletWaasExport }, sdkClient);
		await waasClient.exportPrivateKey({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			displayContainer,
			mfaToken,
			password,
			signedSessionId
		});
	};
	const restoreUserShareForWalletAccount = async ({ walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		await (await getWaasClient()).getWallet({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			signedSessionId,
			walletOperation: _dynamic_labs_wallet_browser_wallet_client.WalletOperation.SIGN_MESSAGE
		});
	};
	const importPrivateKey = async ({ privateKey, thresholdSignatureScheme = _dynamic_labs_wallet_browser_wallet_client.ThresholdSignatureScheme.TWO_OF_TWO }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		await (await getWaasClient()).importPrivateKey({
			authToken: sdkClient.token ?? void 0,
			privateKey,
			signedSessionId,
			thresholdSignatureScheme
		});
	};
	const refreshWalletAccountShares = async ({ password, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		const walletClient = await getWaasClient();
		const mfaToken = await require_getSignedSessionId.consumeMfaTokenIfRequiredForAction({ mfaAction: _dynamic_labs_sdk_api_core.MFAAction.WalletWaasRefresh }, sdkClient);
		return walletClient.refreshWalletAccountShares({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			mfaToken,
			password,
			signedSessionId
		});
	};
	const revokeDelegation = async ({ password, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		return (await getWaasClient()).revokeDelegation({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			password,
			signedSessionId
		});
	};
	const signRawMessage = async ({ message, password, walletAccount }) => {
		if (message.length !== RAW_MESSAGE_MESSAGE_REQUIRED_LENGTH) throw new require_InvalidParamError.InvalidParamError(`Message must be ${RAW_MESSAGE_MESSAGE_REQUIRED_LENGTH} characters long`);
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		const mfaToken = await require_getSignedSessionId.consumeMfaTokenIfRequiredForAction({ mfaAction: _dynamic_labs_sdk_api_core.MFAAction.WalletWaasSign }, sdkClient);
		return { signature: await (await getWaasClient()).signRawMessage({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			message,
			mfaToken,
			password,
			signedSessionId
		}) };
	};
	const terminate = async () => {
		if (!_waasClient) return;
		await (await getWaasClient()).cleanup();
		_waasClient = void 0;
	};
	const updatePassword = async ({ walletAccount, currentPassword, newPassword }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		return (await getWaasClient()).updatePassword({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			existingPassword: currentPassword,
			newPassword,
			signedSessionId
		});
	};
	const signMessage = async ({ message, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		const mfaToken = await require_getSignedSessionId.consumeMfaTokenIfRequiredForAction({ mfaAction: _dynamic_labs_sdk_api_core.MFAAction.WalletWaasSign }, sdkClient);
		return { signature: await (await getWaasClient()).signMessage({
			accountAddress: walletAccount.address,
			authToken: sdkClient.token ?? void 0,
			message,
			mfaToken,
			signedSessionId
		}) };
	};
	const signSerializedTransaction = async ({ serializedTransaction, walletAccount }) => {
		const { signature: signedSessionId } = await require_getSignedSessionId.getSignedSessionId(sdkClient);
		const mfaToken = await require_getSignedSessionId.consumeMfaTokenIfRequiredForAction({ mfaAction: _dynamic_labs_sdk_api_core.MFAAction.WalletWaasSign }, sdkClient);
		return { signature: await (await getWaasClient()).signTransaction({
			authToken: sdkClient.token ?? void 0,
			mfaToken,
			senderAddress: walletAccount.address,
			signedSessionId,
			transaction: serializedTransaction
		}) };
	};
	return {
		backupKeySharesToGoogleDrive,
		createWalletAccount,
		delegateKeyShares,
		exportClientKeyshares,
		exportPrivateKey,
		getWaasClient,
		importPrivateKey,
		refreshWalletAccountShares,
		restoreUserShareForWalletAccount,
		revokeDelegation,
		signMessage,
		signRawMessage,
		signSerializedTransaction,
		terminate,
		updatePassword
	};
};

//#endregion
//#region src/modules/waas/getAllUserWaasAddressesForChain/getAllUserWaasAddressesForChain.ts
const getAllUserWaasAddressesForChain = ({ chain }, client) => {
	if (!client.user) return [];
	return (client.user.verifiedCredentials.filter((credential) => credential.walletProvider === _dynamic_labs_sdk_api_core.WalletProviderEnum.EmbeddedWallet && credential.walletName?.toLowerCase().startsWith(require_constants.DYNAMIC_WAAS_METADATA.normalizedWalletName) && credential.address && credential.chain && require_constants.getChainFromVerifiedCredentialChain(credential.chain) === chain) ?? []).map((credential) => credential.address);
};

//#endregion
//#region src/exports/waasCore.ts
(0, _dynamic_labs_sdk_assert_package_version.assertPackageVersion)(require_constants.name, require_constants.version);

//#endregion
exports.DYNAMIC_WAAS_METADATA = require_constants.DYNAMIC_WAAS_METADATA;
exports.createWaasClient = createWaasClient;
exports.createWaasProvider = createWaasProvider;
exports.getAllUserWaasAddressesForChain = getAllUserWaasAddressesForChain;
exports.getWaasChainNameFromChain = getWaasChainNameFromChain;
