// @dynamic-labs-sdk/client/dist/waas.cjs.js
const require_getNetworkProviderFromNetworkId = require('./getNetworkProviderFromNetworkId-CTKZxd1T.cjs.js');
const require_constants = require('./constants-BbG0v26E.cjs.js');
const require_getVerifiedCredentialForWalletAccount = require('./getVerifiedCredentialForWalletAccount-DAb9lFwa.cjs.js');
const require_NotWaasWalletAccountError = require('./NotWaasWalletAccountError-DUvIdb2w.cjs.js');
const require_filterDuplicates = require('./filterDuplicates-DPNky3ug.cjs.js');
let _dynamic_labs_sdk_assert_package_version = require("@dynamic-labs-sdk/assert-package-version");
let _dynamic_labs_sdk_api_core = require("@dynamic-labs/sdk-api-core");

//#region src/errors/NotWaasWalletProviderError.ts
var NotWaasWalletProviderError = class extends require_constants.BaseError {
	constructor({ walletProviderKey }) {
		super({
			cause: null,
			code: "not_waas_wallet_provider_error",
			docsUrl: null,
			name: "NotWaasWalletProviderError",
			shortMessage: `Wallet provider ${walletProviderKey} is not a Dynamic WaaS wallet provider`
		});
	}
};

//#endregion
//#region src/modules/waas/getWaasWalletProviderFromWalletAccount/getWaasWalletProviderFromWalletAccount.ts
const getWaasWalletProviderFromWalletAccount = ({ walletAccount }, client) => {
	const walletProvider = require_getVerifiedCredentialForWalletAccount.getWalletProviderFromWalletAccount({ walletAccount }, client);
	if (!require_getVerifiedCredentialForWalletAccount.isWaasWalletProvider(walletProvider)) throw new require_NotWaasWalletAccountError.NotWaasWalletAccountError({ walletAddress: walletAccount.address });
	return walletProvider;
};

//#endregion
//#region src/modules/waas/backupWaasKeySharesToGoogleDrive/backupWaasKeySharesToGoogleDrive.ts
/**
* Backs up WaaS wallet key shares to Google Drive for secure recovery.
*
* This function creates a backup of the wallet's cryptographic key shares
* to Google Drive, allowing users to recover their wallet later.
*
* @param params.walletAccount - The WaaS wallet account to backup.
* @param [params.password] - Optional password to encrypt the backup.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the backup process is complete.
*/
const backupWaasKeySharesToGoogleDrive = async ({ password, walletAccount }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderFromWalletAccount({ walletAccount }, client).backupKeySharesToGoogleDrive({
		password,
		walletAccount
	});
};

//#endregion
//#region src/modules/waas/createWaasWalletAccounts/createWaasWalletAccounts.ts
/**
* Creates wallet accounts for the specified chains.
*
* This function creates wallet accounts for the specified chains.
* If the same chain is passed multiple times, just as many wallet accounts will be created
* for that chain.
*
* @example
* ```ts
* // Will create 2 wallet accounts for EVM and 1 for SOL
* await createWaasWalletAccounts({ chains: ['EVM', 'EVM', 'SOL'] });
* ```
*
* @param params.chains - The chains to create wallet accounts for.
* @param params.password - The optional password to use for the wallet accounts.
* @param params.thresholdSignatureScheme - The threshold signature scheme to use for the wallet accounts. (Default: TWO_OF_TWO)
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when all wallet accounts are created.
*/
const createWaasWalletAccounts = async ({ chains, password, thresholdSignatureScheme }, client = require_constants.getDefaultClient()) => {
	const supportedChains = [
		"EVM",
		"SOL",
		"SUI"
	];
	const filteredSupportedChains = chains.filter((chain) => supportedChains.includes(chain));
	if (filteredSupportedChains.length === 0) return;
	const walletProviderKeys = filteredSupportedChains.map((chain) => require_getVerifiedCredentialForWalletAccount.formatWalletProviderKey({
		chain,
		displayName: require_constants.DYNAMIC_WAAS_METADATA.displayName,
		walletProviderType: _dynamic_labs_sdk_api_core.WalletProviderEnum.EmbeddedWallet
	}));
	const walletProviderEntries = require_filterDuplicates.filterDuplicates(walletProviderKeys).map((walletProviderKey) => {
		const walletProvider = require_getVerifiedCredentialForWalletAccount.getWalletProviderByKey({ walletProviderKey }, client);
		if (!walletProvider) throw new require_getVerifiedCredentialForWalletAccount.NoWalletProviderFoundError({ walletProviderKey });
		if (!require_getVerifiedCredentialForWalletAccount.isWaasWalletProvider(walletProvider)) throw new NotWaasWalletProviderError({ walletProviderKey });
		return [walletProviderKey, walletProvider];
	});
	const walletProviderMap = Object.fromEntries(walletProviderEntries);
	await Promise.all(walletProviderKeys.map((walletProviderKey) => walletProviderMap[walletProviderKey].createWalletAccount({
		password,
		thresholdSignatureScheme
	})));
	await require_NotWaasWalletAccountError.refreshAuth(client);
};

//#endregion
//#region src/modules/waas/delegateWaasKeyShares/delegateWaasKeyShares.ts
/**
* Delegates cryptographic key shares for a WaaS wallet account.
*
* This function allows the delegation of key management responsibilities,
* enabling distributed key management for enhanced security.
*
* @param params.walletAccount - The WaaS wallet account to delegate shares for.
* @param [params.password] - Optional password for key share operations.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the key delegation is complete.
*/
const delegateWaasKeyShares = async ({ password, walletAccount }, client = require_constants.getDefaultClient()) => {
	await getWaasWalletProviderFromWalletAccount({ walletAccount }, client).delegateKeyShares({
		password,
		walletAccount
	});
	await require_NotWaasWalletAccountError.refreshAuth(client);
};

//#endregion
//#region src/modules/waas/exportWaasClientKeyshares/exportWaasClientKeyshares.ts
/**
* Exports the client-side key shares for a WaaS wallet account.
*
* This function retrieves the cryptographic key shares that are stored
* on the client side for backup or migration purposes.
*
* @param params.walletAccount - The WaaS wallet account to export shares from.
* @param [params.password] - Optional password to encrypt the exported shares.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves to the exported key shares.
*/
const exportWaasClientKeyshares = async ({ password, walletAccount }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderFromWalletAccount({ walletAccount }, client).exportClientKeyshares({
		password,
		walletAccount
	});
};

//#endregion
//#region src/modules/waas/exportWaasPrivateKey/exportWaasPrivateKey.ts
/**
* Exports the private key for a WaaS wallet account with secure display.
*
* This function reveals the private key for a WaaS wallet and displays it
* securely within an iframe dynamically inserted into the specified HTML container element.
*
* @param params.walletAccount - The WaaS wallet account to export the private key from.
* @param params.displayContainer - The HTML element where the iframe will be injected.
* @param [params.password] - Optional password to decrypt the private key.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the private key export is complete.
*/
const exportWaasPrivateKey = async ({ displayContainer, password, walletAccount }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderFromWalletAccount({ walletAccount }, client).exportPrivateKey({
		displayContainer,
		password,
		walletAccount
	});
};

//#endregion
//#region src/modules/wallets/userHasEmbeddedWalletForChain/userHasEmbeddedWalletForChain.ts
const userHasEmbeddedWalletForChain = ({ user, chain }) => {
	return user.verifiedCredentials.some((vc) => vc.walletProvider === _dynamic_labs_sdk_api_core.WalletProviderEnum.EmbeddedWallet && vc.chain && require_constants.getChainFromVerifiedCredentialChain(vc.chain) === chain);
};

//#endregion
//#region src/modules/waas/isDynamicWaasEnabled/isDynamicWaasEnabled.ts
/**
* Checks if Dynamic WaaS (Wallet as a Service) is enabled for the current project.
*
* This function examines the project settings to determine if Dynamic WaaS is
* enabled in the Dynamic Dashboard or not.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns True if Dynamic WaaS is enabled, false otherwise.
*/
const isDynamicWaasEnabled = (client = require_constants.getDefaultClient()) => {
	const projectSettings = client.projectSettings;
	require_constants.assertDefined(projectSettings, "Project settings are not available");
	return projectSettings.sdk.embeddedWallets?.defaultWalletVersion === _dynamic_labs_sdk_api_core.EmbeddedWalletVersionEnum.V3;
};

//#endregion
//#region src/modules/waas/getChainsMissingWaasWalletAccounts/getChainsMissingWaasWalletAccounts.ts
/**
* Gets all chains missing a waas wallet account for the current user.
*
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns An array of all chains that have waas wallet accounts enabled
* but for which the current user does not yet have an embedded wallet account.
*/
const getChainsMissingWaasWalletAccounts = (client = require_constants.getDefaultClient()) => {
	const projectSettings = client.projectSettings;
	require_constants.assertDefined(projectSettings, "Project settings are not available");
	const embeddedWalletSettings = projectSettings.sdk.embeddedWallets;
	const isWaasEnabled = isDynamicWaasEnabled(client);
	const user = client.user;
	if (!user || !isWaasEnabled) return [];
	return (embeddedWalletSettings?.chainConfigurations?.filter((chain) => chain.enabled) ?? []).map((chain) => chain.name).filter((chain) => !userHasEmbeddedWalletForChain({
		chain,
		user
	}));
};

//#endregion
//#region src/modules/waas/hasDelegatedAccess/hasDelegatedAccess.ts
/**
* Checks if a WaaS wallet account has delegated access.
*
* @param params.walletAccount - The WaaS wallet account to check for delegated access.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns True if the wallet account has delegated access, false otherwise.
*/
const hasDelegatedAccess = ({ walletAccount }, client = require_constants.getDefaultClient()) => {
	const verifiedCredential = require_getVerifiedCredentialForWalletAccount.getVerifiedCredentialForWalletAccount({ walletAccount }, client);
	require_constants.assertDefined(verifiedCredential, "Verified credential not found for Waas wallet account");
	return !!verifiedCredential.walletProperties?.keyShares?.some((keyShare) => keyShare.backupLocation === "delegated");
};

//#endregion
//#region src/modules/waas/getWaasWalletProviderByChain/getWaasWalletProviderByChain.ts
const getWaasWalletProviderByChain = ({ chain }, client) => {
	const waasProvider = require_getVerifiedCredentialForWalletAccount.findWaasWalletProviderByChain({ chain }, client);
	if (!waasProvider) throw new require_getVerifiedCredentialForWalletAccount.NoWalletProviderFoundError({ walletProviderKey: require_getVerifiedCredentialForWalletAccount.formatWalletProviderKey({
		chain,
		displayName: require_constants.DYNAMIC_WAAS_METADATA.displayName,
		walletProviderType: _dynamic_labs_sdk_api_core.WalletProviderEnum.EmbeddedWallet
	}) });
	return waasProvider;
};

//#endregion
//#region src/modules/waas/importWaasPrivateKey/importWaasPrivateKey.ts
/**
* Imports an existing private key to create a new WaaS wallet account.
*
* This function allows users to import their existing private key into
* Dynamic's WaaS infrastructure for secure management.
*
* @param params.chain - The chain for the imported wallet.
* @param params.privateKey - The private key to import.
* @param [params.thresholdSignatureScheme] - Optional threshold signature scheme configuration. Default is TWO_OF_TWO.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the private key is successfully imported.
*/
const importWaasPrivateKey = async ({ chain, privateKey, thresholdSignatureScheme }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderByChain({ chain }, client).importPrivateKey({
		privateKey,
		thresholdSignatureScheme
	});
};

//#endregion
//#region src/modules/waas/isPasswordRequiredForWaasWallets/isPasswordRequiredForWaasWallets.ts
/**
* Checks if the user Waas wallets are password protected.
*
* If the user's Waas wallets are password protected, it means that user's will have to be prompted for a
* password for the first wallet action that happens on each session.
* The password is also required for creating any new Waas wallets.
*
* @param client - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns True if the user has any password protected Waas wallets, false otherwise
*/
const isPasswordRequiredForWaasWallets = (client = require_constants.getDefaultClient()) => {
	const { user } = client;
	require_constants.assertDefined(user, "User is not authenticated");
	return user.verifiedCredentials.some((credential) => credential.walletProperties?.keyShares?.some((keyShare) => keyShare.passwordEncrypted));
};

//#endregion
//#region src/modules/waas/refreshWaasWalletAccountShares/refreshWaasWalletAccountShares.ts
/**
* Refreshes the cryptographic key shares for a WaaS wallet account.
*
* This function generates new key shares for the wallet while maintaining
* the same threshold signature scheme.
*
* @param params.walletAccount - The WaaS wallet account to refresh shares for.
* @param [params.password] - Optional password for key share operations.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the key shares are refreshed.
*/
const refreshWaasWalletAccountShares = async ({ walletAccount, password }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderFromWalletAccount({ walletAccount }, client).refreshWalletAccountShares({
		password,
		walletAccount
	});
};

//#endregion
//#region src/modules/waas/revokeWaasDelegation/revokeWaasDelegation.ts
/**
* Revokes the delegation of a WaaS wallet account.
*
* @param params.walletAccount - The WaaS wallet account to revoke delegation for.
* @param [params.password] - Optional password for delegation operations.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the delegation is revoked.
*/
const revokeWaasDelegation = async ({ password, walletAccount }, client = require_constants.getDefaultClient()) => {
	await getWaasWalletProviderFromWalletAccount({ walletAccount }, client).revokeDelegation({
		password,
		walletAccount
	});
	await require_NotWaasWalletAccountError.refreshAuth(client);
};

//#endregion
//#region src/modules/wallets/userHasExternalWallet/userHasExternalWallet.ts
const userHasExternalWallet = (user) => {
	return user?.verifiedCredentials.some((vc) => vc.format === _dynamic_labs_sdk_api_core.JwtVerifiedCredentialFormatEnum.Blockchain && vc.walletProvider !== _dynamic_labs_sdk_api_core.WalletProviderEnum.EmbeddedWallet);
};

//#endregion
//#region src/modules/waas/shouldAutoCreateWalletForChain/shouldAutoCreateWalletForChain.ts
/**
* Checks if a WaaS wallet should be automatically created for the specified blockchain.
*
* This function determines whether the current user configuration and project
* settings require automatic wallet creation for the given chain, considering
* whether the user already has an embedded wallet for the enabled chains or an external wallet.
*
* @param params.chain - The blockchain network to check.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns True if a wallet should be auto-created for the chain, false otherwise.
*/
const shouldAutoCreateWalletForChain = ({ chain }, client = require_constants.getDefaultClient()) => {
	require_constants.assertDefined(client.projectSettings, "Project settings are not available");
	const embeddedWalletSettings = client.projectSettings.sdk.embeddedWallets;
	if (!embeddedWalletSettings?.automaticEmbeddedWalletCreation || !client.user) return false;
	const automaticEmbeddedWalletCreationForExternalEnabled = embeddedWalletSettings?.automaticEmbeddedWalletCreationForExternal;
	if (!(!userHasExternalWallet(client.user) || automaticEmbeddedWalletCreationForExternalEnabled)) return false;
	return getChainsMissingWaasWalletAccounts(client).includes(chain);
};

//#endregion
//#region src/modules/waas/updateWaasPassword/updateWaasPassword.ts
/**
* Updates the password for a WaaS wallet account.
*
* This function changes the password used to encrypt and protect the
* cryptographic key shares for the specified wallet account.
*
* @param params.walletAccount - The WaaS wallet account to update the password for.
* @param params.currentPassword - The current password for the wallet.
* @param params.newPassword - The new password to set for the wallet.
* @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
* @returns A promise that resolves when the password is successfully updated.
*/
const updateWaasPassword = async ({ currentPassword, newPassword, walletAccount }, client = require_constants.getDefaultClient()) => {
	return getWaasWalletProviderFromWalletAccount({ walletAccount }, client).updatePassword({
		currentPassword,
		newPassword,
		walletAccount
	});
};

//#endregion
//#region src/exports/waas.ts
(0, _dynamic_labs_sdk_assert_package_version.assertPackageVersion)(require_constants.name, require_constants.version);

//#endregion
exports.NotWaasWalletProviderError = NotWaasWalletProviderError;
exports.backupWaasKeySharesToGoogleDrive = backupWaasKeySharesToGoogleDrive;
exports.createWaasWalletAccounts = createWaasWalletAccounts;
exports.delegateWaasKeyShares = delegateWaasKeyShares;
exports.exportWaasClientKeyshares = exportWaasClientKeyshares;
exports.exportWaasPrivateKey = exportWaasPrivateKey;
exports.getChainsMissingWaasWalletAccounts = getChainsMissingWaasWalletAccounts;
exports.hasDelegatedAccess = hasDelegatedAccess;
exports.importWaasPrivateKey = importWaasPrivateKey;
exports.isDynamicWaasEnabled = isDynamicWaasEnabled;
exports.isPasswordRequiredForWaasWallets = isPasswordRequiredForWaasWallets;
exports.isWaasWalletAccount = require_getVerifiedCredentialForWalletAccount.isWaasWalletAccount;
exports.refreshWaasWalletAccountShares = refreshWaasWalletAccountShares;
exports.revokeWaasDelegation = revokeWaasDelegation;
exports.shouldAutoCreateWalletForChain = shouldAutoCreateWalletForChain;
exports.updateWaasPassword = updateWaasPassword;
