// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useAuthenticatePasskeyMFA/useAuthenticatePasskeyMFA.js
'use client'
import { __awaiter } from '../../../../../_virtual/_tslib.js';
import { useCallback } from 'react';
import { authenticatePasskeyMFA } from '@dynamic-labs-sdk/client';
import { dynamicEvents } from '../../../events/dynamicEvents.js';
import { useUserAuth } from '../useUserAuth/useUserAuth.js';

/**
 * Authenticate passkey for MFA
 *
 * @returns Function to authenticate with a passkey for MFA
 *
 * @example
 * ```tsx
 * const App = () => {
 *   const authenticatePasskeyMFA = useAuthenticatePasskeyMFA();
 *
 *   // Optional: creates a single-use MFA token
 *   const createMfaToken = { singleUse: true };
 *
 *   return (
 *     <button
 *       onClick={() => authenticatePasskeyMFA({ createMfaToken })}
 *     >
 *       Authenticate passkey
 *     </button>
 *  );
 * }
 */
const useAuthenticatePasskeyMFA = () => {
    const { completeAuth } = useUserAuth({});
    return useCallback((props) => __awaiter(void 0, void 0, void 0, function* () {
        let mfaToken = undefined;
        yield completeAuth({
            updateJwtFunction: () => __awaiter(void 0, void 0, void 0, function* () {
                const response = yield authenticatePasskeyMFA(props);
                // eslint-disable-next-line prefer-destructuring
                mfaToken = response.mfaToken;
                dynamicEvents.emit('mfaCompletionSuccess', {
                    mfaToken: response.mfaToken,
                });
                return {
                    isEmailVerificationRequired: false,
                    isSmsVerificationRequired: false,
                    missingFields: [],
                    // VerifyResponse is compatible with UpdateSelfResponse for MFA flows:
                    // Both types have: expiresAt, jwt, minifiedJwt, user (used in useUserAuth.ts:329-332)
                    updateUserProfileResponse: response,
                };
            }),
        });
        return mfaToken;
    }), [completeAuth]);
};

export { useAuthenticatePasskeyMFA };
