// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useAuthenticatePasskeyMFA/useAuthenticatePasskeyMFA.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../../_virtual/_tslib.cjs');
var React = require('react');
var client = require('@dynamic-labs-sdk/client');
var dynamicEvents = require('../../../events/dynamicEvents.cjs');
var useUserAuth = require('../useUserAuth/useUserAuth.cjs');

/**
 * Authenticate passkey for MFA
 *
 * @returns Function to authenticate with a passkey for MFA
 *
 * @example
 * ```tsx
 * const App = () => {
 *   const authenticatePasskeyMFA = useAuthenticatePasskeyMFA();
 *
 *   // Optional: creates a single-use MFA token
 *   const createMfaToken = { singleUse: true };
 *
 *   return (
 *     <button
 *       onClick={() => authenticatePasskeyMFA({ createMfaToken })}
 *     >
 *       Authenticate passkey
 *     </button>
 *  );
 * }
 */
const useAuthenticatePasskeyMFA = () => {
    const { completeAuth } = useUserAuth.useUserAuth({});
    return React.useCallback((props) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        let mfaToken = undefined;
        yield completeAuth({
            updateJwtFunction: () => _tslib.__awaiter(void 0, void 0, void 0, function* () {
                const response = yield client.authenticatePasskeyMFA(props);
                // eslint-disable-next-line prefer-destructuring
                mfaToken = response.mfaToken;
                dynamicEvents.dynamicEvents.emit('mfaCompletionSuccess', {
                    mfaToken: response.mfaToken,
                });
                return {
                    isEmailVerificationRequired: false,
                    isSmsVerificationRequired: false,
                    missingFields: [],
                    // VerifyResponse is compatible with UpdateSelfResponse for MFA flows:
                    // Both types have: expiresAt, jwt, minifiedJwt, user (used in useUserAuth.ts:329-332)
                    updateUserProfileResponse: response,
                };
            }),
        });
        return mfaToken;
    }), [completeAuth]);
};

exports.useAuthenticatePasskeyMFA = useAuthenticatePasskeyMFA;
