// @dynamic-labs/utils/src/services/PlatformService/createBrowserPlatformService/createBrowserPlatformService.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../_virtual/_tslib.cjs');
var tldts = require('tldts');

const createBrowserPlatformService = (window) => ({
    downloadFile: (fileName, file) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const element = document.createElement('a');
        element.href = URL.createObjectURL(file);
        element.download = fileName;
        document.body.appendChild(element);
        element.click();
        document.body.removeChild(element);
        URL.revokeObjectURL(element.href);
    }),
    getDisplayOrigin: () => window.location.origin,
    getHost: () => window.location.host,
    getHostname: () => window.location.hostname,
    getOrigin: () => window.location.origin,
    getTLD: (domain) => {
        // Passing the allowPrivateDomains option prevents returning the actual TLD
        // for domains that have delegated subdomains like herokuapp.com or s3.amazonaws.com
        // full list is contained here https://publicsuffix.org/list/effective_tld_names.dat
        // separated by ICANN DOMAINS and PRIVATE DOMAINS
        // so for instance parse('someapp.herokuapp.com') will return 'someapp.herokuapp.com' as the domain
        // whereas parse('app.dynamic.xyz') will return 'dynamic.xyz'
        const data = tldts.parse(domain || window.location.hostname, {
            allowPrivateDomains: true,
        });
        return data.domain || undefined;
    },
    getUrl: () => new URL(window.location.href),
    isNativeMobile: false,
    openURL: (url_1, ...args_1) => _tslib.__awaiter(void 0, [url_1, ...args_1], void 0, function* (url, target = 'self', features = '') {
        if (target === 'blank') {
            window.open(url, '_blank', features);
        }
        else {
            window.location.assign(url);
        }
    }),
});

exports.createBrowserPlatformService = createBrowserPlatformService;
