// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampFeeType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * The type of fee.
 * @export
 * @enum {string}
 */
exports.CoinbaseOnrampFeeType = void 0;
(function (CoinbaseOnrampFeeType) {
    CoinbaseOnrampFeeType["Network"] = "FEE_TYPE_NETWORK";
    CoinbaseOnrampFeeType["Exchange"] = "FEE_TYPE_EXCHANGE";
})(exports.CoinbaseOnrampFeeType || (exports.CoinbaseOnrampFeeType = {}));
function CoinbaseOnrampFeeTypeFromJSON(json) {
    return CoinbaseOnrampFeeTypeFromJSONTyped(json);
}
function CoinbaseOnrampFeeTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampFeeTypeToJSON(value) {
    return value;
}

exports.CoinbaseOnrampFeeTypeFromJSON = CoinbaseOnrampFeeTypeFromJSON;
exports.CoinbaseOnrampFeeTypeFromJSONTyped = CoinbaseOnrampFeeTypeFromJSONTyped;
exports.CoinbaseOnrampFeeTypeToJSON = CoinbaseOnrampFeeTypeToJSON;
