// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampOrderStatus.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * The status of an onramp order.
 * @export
 * @enum {string}
 */
exports.CoinbaseOnrampOrderStatus = void 0;
(function (CoinbaseOnrampOrderStatus) {
    CoinbaseOnrampOrderStatus["PendingAuth"] = "ONRAMP_ORDER_STATUS_PENDING_AUTH";
    CoinbaseOnrampOrderStatus["PendingPayment"] = "ONRAMP_ORDER_STATUS_PENDING_PAYMENT";
    CoinbaseOnrampOrderStatus["Processing"] = "ONRAMP_ORDER_STATUS_PROCESSING";
    CoinbaseOnrampOrderStatus["Completed"] = "ONRAMP_ORDER_STATUS_COMPLETED";
    CoinbaseOnrampOrderStatus["Failed"] = "ONRAMP_ORDER_STATUS_FAILED";
})(exports.CoinbaseOnrampOrderStatus || (exports.CoinbaseOnrampOrderStatus = {}));
function CoinbaseOnrampOrderStatusFromJSON(json) {
    return CoinbaseOnrampOrderStatusFromJSONTyped(json);
}
function CoinbaseOnrampOrderStatusFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampOrderStatusToJSON(value) {
    return value;
}

exports.CoinbaseOnrampOrderStatusFromJSON = CoinbaseOnrampOrderStatusFromJSON;
exports.CoinbaseOnrampOrderStatusFromJSONTyped = CoinbaseOnrampOrderStatusFromJSONTyped;
exports.CoinbaseOnrampOrderStatusToJSON = CoinbaseOnrampOrderStatusToJSON;
