// @dynamic-labs/sdk-api-core/src/models/RoomTypeEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.RoomTypeEnum = void 0;
(function (RoomTypeEnum) {
    RoomTypeEnum["Threshold"] = "threshold";
    RoomTypeEnum["Full"] = "full";
})(exports.RoomTypeEnum || (exports.RoomTypeEnum = {}));
function RoomTypeEnumFromJSON(json) {
    return RoomTypeEnumFromJSONTyped(json);
}
function RoomTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function RoomTypeEnumToJSON(value) {
    return value;
}

exports.RoomTypeEnumFromJSON = RoomTypeEnumFromJSON;
exports.RoomTypeEnumFromJSONTyped = RoomTypeEnumFromJSONTyped;
exports.RoomTypeEnumToJSON = RoomTypeEnumToJSON;
