// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampOrderPaymentLinkType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * The type of payment link.
 * @export
 * @enum {string}
 */
exports.CoinbaseOnrampOrderPaymentLinkType = void 0;
(function (CoinbaseOnrampOrderPaymentLinkType) {
    CoinbaseOnrampOrderPaymentLinkType["PaymentLinkTypeApplePayButton"] = "PAYMENT_LINK_TYPE_APPLE_PAY_BUTTON";
})(exports.CoinbaseOnrampOrderPaymentLinkType || (exports.CoinbaseOnrampOrderPaymentLinkType = {}));
function CoinbaseOnrampOrderPaymentLinkTypeFromJSON(json) {
    return CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped(json);
}
function CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampOrderPaymentLinkTypeToJSON(value) {
    return value;
}

exports.CoinbaseOnrampOrderPaymentLinkTypeFromJSON = CoinbaseOnrampOrderPaymentLinkTypeFromJSON;
exports.CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped = CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped;
exports.CoinbaseOnrampOrderPaymentLinkTypeToJSON = CoinbaseOnrampOrderPaymentLinkTypeToJSON;
