// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampOrderPaymentMethod.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * The type of payment method to be used to complete an onramp order.
 * @export
 * @enum {string}
 */
var CoinbaseOnrampOrderPaymentMethod;
(function (CoinbaseOnrampOrderPaymentMethod) {
    CoinbaseOnrampOrderPaymentMethod["GuestCheckoutApplePay"] = "GUEST_CHECKOUT_APPLE_PAY";
})(CoinbaseOnrampOrderPaymentMethod || (CoinbaseOnrampOrderPaymentMethod = {}));
function CoinbaseOnrampOrderPaymentMethodFromJSON(json) {
    return CoinbaseOnrampOrderPaymentMethodFromJSONTyped(json);
}
function CoinbaseOnrampOrderPaymentMethodFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampOrderPaymentMethodToJSON(value) {
    return value;
}

export { CoinbaseOnrampOrderPaymentMethod, CoinbaseOnrampOrderPaymentMethodFromJSON, CoinbaseOnrampOrderPaymentMethodFromJSONTyped, CoinbaseOnrampOrderPaymentMethodToJSON };
