// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampBuyUrlExperience.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Default visual experience. Either transfer funds from Coinbase (send) or buy assets (buy)
 * @export
 * @enum {string}
 */
var CoinbaseOnrampBuyUrlExperience;
(function (CoinbaseOnrampBuyUrlExperience) {
    CoinbaseOnrampBuyUrlExperience["Buy"] = "buy";
    CoinbaseOnrampBuyUrlExperience["Send"] = "send";
})(CoinbaseOnrampBuyUrlExperience || (CoinbaseOnrampBuyUrlExperience = {}));
function CoinbaseOnrampBuyUrlExperienceFromJSON(json) {
    return CoinbaseOnrampBuyUrlExperienceFromJSONTyped(json);
}
function CoinbaseOnrampBuyUrlExperienceFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampBuyUrlExperienceToJSON(value) {
    return value;
}

export { CoinbaseOnrampBuyUrlExperience, CoinbaseOnrampBuyUrlExperienceFromJSON, CoinbaseOnrampBuyUrlExperienceFromJSONTyped, CoinbaseOnrampBuyUrlExperienceToJSON };
