// @dynamic-labs/sdk-api-core/src/models/WalletTransactionType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.WalletTransactionType = void 0;
(function (WalletTransactionType) {
    WalletTransactionType["Sent"] = "sent";
    WalletTransactionType["Receive"] = "receive";
    WalletTransactionType["Swap"] = "swap";
})(exports.WalletTransactionType || (exports.WalletTransactionType = {}));
function WalletTransactionTypeFromJSON(json) {
    return WalletTransactionTypeFromJSONTyped(json);
}
function WalletTransactionTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function WalletTransactionTypeToJSON(value) {
    return value;
}

exports.WalletTransactionTypeFromJSON = WalletTransactionTypeFromJSON;
exports.WalletTransactionTypeFromJSONTyped = WalletTransactionTypeFromJSONTyped;
exports.WalletTransactionTypeToJSON = WalletTransactionTypeToJSON;
