// @dynamic-labs/solana/src/injected/BackpackSol/BackpackSol.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.cjs');
var utils = require('@dynamic-labs/utils');
var walletBook = require('@dynamic-labs/wallet-book');
var InjectedWalletBase = require('../InjectedWalletBase/InjectedWalletBase.cjs');

class BackpackSol extends InjectedWalletBase.InjectedWalletBase {
    constructor(props) {
        super(props);
        this.name = 'Backpack';
        this.overrideKey = 'backpacksol';
        this.walletConnectWalletBookEntry = walletBook.findWalletBookWallet(props.walletBook, this.key);
    }
    getSigner() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            return (_a = this.solProviderHelper) === null || _a === void 0 ? void 0 : _a.connect();
        });
    }
    signMessage(messageToSign) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const publicAddress = yield this.getAddress();
            if (!publicAddress) {
                return undefined;
            }
            const provider = yield this.getSigner();
            if (!provider) {
                return undefined;
            }
            const signedMessage = yield provider.signMessage(Buffer.from(messageToSign, 'utf8'));
            if (!signedMessage) {
                return undefined;
            }
            if (typeof signedMessage === 'object' && 'signature' in signedMessage) {
                return utils.bufferToBase64(signedMessage.signature);
            }
            return utils.bufferToBase64(signedMessage);
        });
    }
}
Object.defineProperty(BackpackSol, 'key', {
    value: 'backpacksol',
    writable: false,
});

exports.BackpackSol = BackpackSol;
