// @dynamic-labs/wallet-connect/src/getSignClientSingleton/getSignClientSingleton.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.cjs');
var SignClient = require('@walletconnect/sign-client');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var SignClient__default = /*#__PURE__*/_interopDefaultLegacy(SignClient);

/**
 * WalletConnect's SignClient would already internally share its core if we had
 * multiple instances of it, so instead we will just use a singleton already.
 *
 * A SignClient is able to handle multiple sessions and multi chain connections.
 */
let signClientSingleton = undefined;
const getSignClientSingleton = (_a) => _tslib.__awaiter(void 0, [_a], void 0, function* ({ projectId, appIcon, appName, }) {
    if (!signClientSingleton) {
        signClientSingleton = yield SignClient__default["default"].init({
            customStoragePrefix: `dynamicauth_${projectId}_walletconnect`,
            metadata: {
                description: '',
                icons: [appIcon],
                name: appName,
                url: '',
            },
            projectId,
        });
    }
    return signClientSingleton;
});

exports.getSignClientSingleton = getSignClientSingleton;
