// @dynamic-labs/sdk-api-core/src/models/CoinbaseOnrampOrderPaymentLinkType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * The type of payment link.
 * @export
 * @enum {string}
 */
var CoinbaseOnrampOrderPaymentLinkType;
(function (CoinbaseOnrampOrderPaymentLinkType) {
    CoinbaseOnrampOrderPaymentLinkType["PaymentLinkTypeApplePayButton"] = "PAYMENT_LINK_TYPE_APPLE_PAY_BUTTON";
})(CoinbaseOnrampOrderPaymentLinkType || (CoinbaseOnrampOrderPaymentLinkType = {}));
function CoinbaseOnrampOrderPaymentLinkTypeFromJSON(json) {
    return CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped(json);
}
function CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CoinbaseOnrampOrderPaymentLinkTypeToJSON(value) {
    return value;
}

export { CoinbaseOnrampOrderPaymentLinkType, CoinbaseOnrampOrderPaymentLinkTypeFromJSON, CoinbaseOnrampOrderPaymentLinkTypeFromJSONTyped, CoinbaseOnrampOrderPaymentLinkTypeToJSON };
